### PR TITLE
Scalafix rules to rewrite to new ServerBuilder API automatically

### PR DESCRIPTION
**Pull request by jrudolph**
Still TODO:
 * [x] put them into correct packages
 * [x] some FIXMEs
 * [x] documentation + inclusion in the release notes

### CODE REVIEW

**Comment by ennru** [comment]
Let's make sure Scala Steward can pick these up so that it can propose the rewrites in its PRs.

**Comment by jrudolph** [comment]
> Let's make sure Scala Steward can pick these up so that it can propose the rewrites in its PRs.

I had it on the list but anyway thanks for the reminder, opened PR here: https://github.com/scala-steward-org/scala-steward/pull/1562

**Comment by jrudolph** [comment]
Tested the rules on a separate project and added docs section about how to run it in release notes and migration notes.

**Comment by jrudolph** [comment]
While testing out required changes for akka-grpc I found one remaining issue:

The new ServerBuilder currently doesn't allow specifying a custom materializer and the scalafix rule will produce broken code. Should we do a last minute fix to add materializer customization to the ServerBuilders and also update the fix rule for those occurrences?

So far, we haven't completely decided if we want to get rid of custom materializers completely or if we want to keep supporting them (which would technically be fine right now).

**Comment by jrudolph** [comment]
> The new ServerBuilder currently doesn't allow specifying a custom materializer and the scalafix rule will produce broken code. Should we do a last minute fix to add materializer customization to the ServerBuilders and also update the fix rule for those occurrences?
> 
> So far, we haven't completely decided if we want to get rid of custom materializers completely or if we want to keep supporting them (which would technically be fine right now).

Fixed now in #3414. I now also changed the code to include this case.

**Comment by jrudolph** [comment]
Now finished from my side. See 

https://github.com/akka/akka-http/pull/3407/files#diff-d4ea2f20df23bdbfd1e6beedb0020f9aR38-R72

and

https://github.com/akka/akka-http/pull/3407/files#diff-2c37bf7912231bf47cbecc4c9ab209ccR34-R56

for all supported automatic rewrites.

File: akka-http-scalafix/README.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+ * Add before/after test file in scalafix-test-input / scalafix-test-output
+ * Add rule in scalafix-rules
+ * run test in `akka-http-scalafix-tests`
+
+## Applying locally defined rules to docs examples
+
+ * run `scalafixEnable` on the sbt shell (this will unfortunately require a complete rebuild afterwards)
+ * run `set scalacOptions in ThisBuild += "-P:semanticdb:synthetics:on"` to allow access to synthetics
+ * e.g. run docs/scalafixAll MigrateToServerBuilder`
```
Comment (ennru) [inline akka-http-scalafix/README.md:13]:
```suggestion
 * e.g. run `docs/scalafixAll MigrateToServerBuilder`
```

File: akka-http-scalafix/scalafix-rules/src/main/scala/akka/http/fix/MigrateToServerBuilder.scala [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+          require(sig.parameterLists(1)(0).signature.toString == "Materializer")
+          (t.parent.get, t.parent.get.asInstanceOf[Term.Apply].args.head)
+        }.toOption
+
+      val materializerLint: Option[Patch] = materializerAndTarget.map {
+        case (_, matArg) =>
+          Patch.lint(Diagnostic(
+            "custom-materializer-warning",
+            "Custom materializers are often not needed any more. You can often just remove custom materializers and " +
+              "just use the system materializer which is supplied automatically.",
```
Comment (ennru) [inline akka-http-scalafix/scalafix-rules/src/main/scala/akka/http/fix/MigrateToServerBuilder.scala:30]:
```suggestion
            "Custom materializers are often not needed any more. You can often remove custom materializers and " +
              "use the system materializer which is supplied automatically.",
```

File: akka-http-scalafix/scalafix-rules/src/main/scala/akka/http/fix/MigrateToServerBuilder.scala [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+      patchTree(targetTree, http, argExps, targetMethod(argExps("handler"))) + materializerLint
+    }
+
+    def patchTree(t: Tree, http: Term, argExps: Map[String, Term], targetMethod: String): Patch =
+      Patch.replaceTree(t, s"${builder(http, argExps)}.$targetMethod(${argExps("handler")})")
+
+    def handlerIsRoute(handler: Term): Boolean =
+      handler.symbol.info.exists(_.signature.toString contains "Route") || // doesn't seem to work with synthetics on
+        (handler.synthetics match { // only works with `scalacOptions += "-P:semanticdb:synthetics:on"`
+          case ApplyTree(fun, _) :: Nil => fun.symbol.exists(_.displayName == "routeToFlow") /* FIXME: could use real symbol instead */
```
Comment (ennru) [inline akka-http-scalafix/scalafix-rules/src/main/scala/akka/http/fix/MigrateToServerBuilder.scala:48]: Want to clear the FIXME?

File: akka-http-scalafix/scalafix-test-input/src/main/scala/akka/http/fix/MigrateToServerBuilderTest.scala [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  def handler: HttpRequest => Future[HttpResponse] = ???
+  def syncHandler: HttpRequest => HttpResponse = ???
+  def flow: Flow[HttpRequest, HttpResponse, Any] = ???
+  def route: Route = ???
+  trait ServiceRoutes {
+    def route: Route = ???
+  }
+  def service: ServiceRoutes = ???
+
+  // fix: materializer or system explicitly given, see https://github.com/akka/akka-http/issues/3410
```
Comment (ennru) [inline akka-http-scalafix/scalafix-test-input/src/main/scala/akka/http/fix/MigrateToServerBuilderTest.scala:36]: Can be removed.

File: akka-http-scalafix/scalafix-rules/src/main/scala/akka/http/fix/MigrateToServerBuilder.scala (hunk 3 above)
Comment (jrudolph) [reply akka-http-scalafix/scalafix-rules/src/main/scala/akka/http/fix/MigrateToServerBuilder.scala:48]: Clarifying the comment instead:

```suggestion
          case ApplyTree(fun, _) :: Nil => fun.symbol.exists(_.displayName == "routeToFlow") // somewhat inaccurate, but that name should be unique enough for our purposes
```

File: akka-http-scalafix/scalafix-test-input/src/main/scala/akka/http/fix/MigrateToServerBuilderTest.scala (hunk 4 above)
Comment (jrudolph) [inline akka-http-scalafix/scalafix-test-input/src/main/scala/akka/http/fix/MigrateToServerBuilderTest.scala:36]:
```suggestion
```

File: akka-http-scalafix/scalafix-test-output/src/main/scala/akka/http/fix/MigrateToServerBuilderTest.scala [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+  def handler: HttpRequest => Future[HttpResponse] = ???
+  def syncHandler: HttpRequest => HttpResponse = ???
+  def flow: Flow[HttpRequest, HttpResponse, Any] = ???
+  def route: Route = ???
+  trait ServiceRoutes {
+    def route: Route = ???
+  }
+  def service: ServiceRoutes = ???
+
+  // fix: materializer or system explicitly given, see https://github.com/akka/akka-http/issues/3410
```
Comment (jrudolph) [inline akka-http-scalafix/scalafix-test-output/src/main/scala/akka/http/fix/MigrateToServerBuilderTest.scala:32]:
```suggestion
```

**Pull request merged**